Upstream-Status: Inappropriate [OpenXT]

################################################################################
SHORT DESCRIPTION: 
################################################################################
Port git://xenbits.xen.org/xen.git:4da31be1202e9d82f4a0c5b7cb79ec36a941598c

################################################################################
LONG DESCRIPTION: 
################################################################################
commit 4da31be1202e9d82f4a0c5b7cb79ec36a941598c
Author: Frediano Ziglio <frediano.ziglio@citrix.com>
Date:   Mon May 7 13:38:23 2012 +0100

    vgabios: Reduce stack usage getting mode informations
    
    Informations are stored in a structure that is smaller than final one.
    Previous code copy this structure to stack extending with zeroes then
    update it and copy to caller while now the not-extended version is
    copied into stack and then is extended during copy reducing stack
    usage.
    
    Signed-off-by: Frediano Ziglio <frediano.ziglio@citrix.com>
    Committed-by: Keir Fraser <keir@xen.org>

################################################################################
CHANGELOG 
################################################################################
Ported to VGABIOS: Eric Chanudet, chanudete@ainfosec.com, 01/03/2015

################################################################################
REMOVAL 
################################################################################
This patch can be removed if a new version of Bochs VGABIOS is released with
this fix.

################################################################################
UPSTREAM PLAN 
################################################################################
This is already upstream in Xen repository.

################################################################################
INTERNAL DEPENDENCIES 
################################################################################
None.

################################################################################
PATCHES
################################################################################
--- a/vbe.c
+++ b/vbe.c
@@ -1256,10 +1256,10 @@ Bit16u *AX;Bit16u CX; Bit16u ES;Bit16u D
         // error by default is 0x014f which means supported but error
         Bit16u                 result=0x014f;
         Bit16u            ss=get_SS();
-        ModeInfoBlock     info;
         ModeInfoListItem  *cur_info;
         Boolean           using_lfb;
         Bit16u            lfb_addr;
+        ModeInfoBlockCompact   info;
         Bit16u            support_bank_gr_32k;
 
 #ifdef DEBUG
@@ -1277,7 +1277,6 @@ Bit16u *AX;Bit16u CX; Bit16u ES;Bit16u D
 #ifdef DEBUG
                 printf("VBE found mode %x\n",CX);
 #endif
-                memsetb(ss, &info, 0, sizeof(ModeInfoBlock));
                 memcpyb(ss, &info, 0xc000, &(cur_info->info), sizeof(ModeInfoBlockCompact));
                 if (using_lfb) {
                   info.NumberOfBanks = 1;
@@ -1300,6 +1299,10 @@ Bit16u *AX;Bit16u CX; Bit16u ES;Bit16u D
                 }
 
                 result = 0x4f;
+
+                // copy updates in mode_info_block back
+                memsetb(ES, DI, 0, sizeof(ModeInfoBlock));
+                memcpyb(ES, DI, ss, &info, sizeof(info));
         }
         else
         {
@@ -1308,12 +1311,6 @@ Bit16u *AX;Bit16u CX; Bit16u ES;Bit16u D
 #endif
         }
 
-        if (result == 0x4f)
-        {
-                // copy updates in mode_info_block back
-                memcpyb(ES, DI, ss, &info, sizeof(info));
-        }
-
         write_word(ss, AX, result);
 }
 
